src/runtime/methodbinder.cs

```diff
@@ -335,6 +356,55 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
                     continue;
                 }
 
+                var matchedMethod = new MatchedMethod(kwargsMatched, defaultsNeeded, margs, outs, mi);
+                argMatchedMethods.Add(matchedMethod);
+            }
+            if (argMatchedMethods.Count() > 0)
+            {
+                // Order matched methods by number of kwargs matched and get the max possible number
+                // of kwargs matched
+                var bestKwargMatchCount = argMatchedMethods.OrderBy((x) => x.KwargsMatched).Reverse().ToArray()[0].KwargsMatched;
+
+                List<MatchedMethod> bestKwargMatches = new List<MatchedMethod>(argMatchedMethods.Count());
+                foreach (MatchedMethod testMatch in argMatchedMethods)
+                {
+                    if (testMatch.KwargsMatched == bestKwargMatchCount)
+                    {
+                        bestKwargMatches.Add(testMatch);
+                    }
+                }
+
+                // Order by the number of defaults required and find the smallest
+                var fewestDefaultsRequired  = bestKwargMatches.OrderBy((x) => x.DefaultsNeeded).ToArray()[0].DefaultsNeeded;
+
+                List<MatchedMethod> bestDefaultsMatches = new List<MatchedMethod>(bestKwargMatches.Count());
+                foreach (MatchedMethod testMatch in bestKwargMatches)
+                {
+                    if (testMatch.DefaultsNeeded == fewestDefaultsRequired)
+                    {
+                        bestDefaultsMatches.Add(testMatch);
+                    }
+                }
+
+                if (bestDefaultsMatches.Count() > 1 && fewestDefaultsRequired > 0)
+                {
+                    // Best effort for determining method to match on gives multiple possible
+                    // matches and we need at least one default argument - bail from this point
+                    return null;
+                }
+
+                // If we're here either:
+                //      (a) There is only one best match
+                //      (b) There are multiple best matches but none of them require
+                //          default arguments
+                // in the case of (a) we're done by default. For (b) regardless of which
+                // method we choose, all arguments are specified _and_ can be converted
+                // from python to C# so picking any will suffice
+                MatchedMethod bestMatch = bestDefaultsMatches.ToArray()[0];
```

Comment:
If you only ever need the first item from the list, why fill an entire list in the first place?